e2e/stack.go

```diff
@@ -0,0 +1,213 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	tmdb "github.com/cometbft/cometbft-db"
+	opgenesis "github.com/ethereum-optimism/optimism/op-chain-ops/genesis"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/polymerdao/monomer"
+	"github.com/polymerdao/monomer/genesis"
+	"github.com/polymerdao/monomer/node"
+	"github.com/polymerdao/monomer/testutil/testapp"
+	"github.com/polymerdao/monomer/utils"
+	rolluptypes "github.com/polymerdao/monomer/x/rollup/types"
+	"github.com/sourcegraph/conc"
+)
+
+type EventListener interface {
+	OPEventListener
+
+	OnCmdStart(programName string, stdout, stderr io.Reader)
+	OnCmdStopped(programName string, err error)
+}
+
+type Stack struct {
+	l1Address        string
+	monomerAddress   string
+	opNodeAddress    string
+	contractsRootDir string
+	eventListener    EventListener
+	l1BlockTime      time.Duration
+}
+
+// New assumes all ports are available and that all paths exist and are valid.
+func New(l1Address, monomerAddress, opNodeAddress, contractsRootDir string, l1BlockTime time.Duration, eventListener EventListener) *Stack {
+	return &Stack{
+		l1Address:        l1Address,
+		monomerAddress:   monomerAddress,
+		opNodeAddress:    opNodeAddress,
+		contractsRootDir: contractsRootDir,
+		eventListener:    eventListener,
+		l1BlockTime:      l1BlockTime,
+	}
+}
+
+func (s *Stack) Run(parentCtx context.Context) (err error) {
+	var wg conc.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := context.WithCancelCause(parentCtx)
+	defer func() {
+		cancel(err)
+		err = utils.Cause(ctx)
+	}()
+
+	const numConnectAttempts = 20
+
+	// Run anvil.
+	_, l1Port, err := net.SplitHostPort(s.l1Address)
+	if err != nil {
+		return fmt.Errorf("split l1 address into host and port: %v", s.l1Address)
+	}
+	wg.Go(func() {
+		cancel(fmt.Errorf("run cmd: %v", s.runCmd(
+			ctx,
+			"anvil",
+			"--port", l1Port,
+			"--order", "fifo",
+			"--disable-block-gas-limit",
+			"--gas-price", "0",
+			"--block-time", fmt.Sprint(s.l1BlockTime.Seconds()),
+		)))
+	})
+	if !utils.AddressIsReady(s.l1Address, numConnectAttempts) {
+		return fmt.Errorf("could not connect to anvil at address %s", s.l1Address)
+	}
+
+	// Fund an account.
+	anvilHTTPAddress := "http://" + s.l1Address
```

Comment:
There are a bunch of instances of this rpc client / dialContext through `stack` and `op_stack`.

Should they instead live inside of, eg, `NewAnvil(address string)` which could return an error? Might make the "main execution paths" a bit less noisy and easier to read. Similar for other client constructors.

Reply:
Good point. I don't like the address handling either. I'll think about how to improve it.

I didn't make `NewAnvil` take the `address` since that would require it to create an rpc client internally, violating dependency injection. That might be a bit pedantic on my part though.

Reply:
Fixed this by wrapping `url.URL`.